make/cpp2asm_support.cmake

```diff
@@ -132,7 +132,7 @@ endif ()
 ##################################################
 # Preprocessor location and flags
 
-if (UNIX)
+if (UNIX OR (WIN32 AND "${CMAKE_GENERATOR}" MATCHES "MSYS Makefiles"))
```

Comment:
One problem with only looking one line at a time is not knowing how many patterns like this will be needed.  If there's only one, maybe this condition is ok, but certainly if there are many, and IMHO for this one too, it is better to generalize.  We don't want to have to come back later and update this for "MinGW Makefiles" too.  Better to use some general variable which captures what is happening here: this code is not dependent on the OS but on the compiler.  So instead of mentioning UNIX or WIN32 it should check something more like CMAKE_COMPILER_IS_CLANG and CMAKE_COMPILER_IS_GNUCC.  I would make a new variable named CMAKE_COMPILER_IS_GNUCC_COMPATIBLE or something, set it for both _GNUCC and _CLANG, and use it here.  I would then go through the rest of the cmake code and replace similar uses of UNIX or WIN32 with the new variable.

Reply:
Maybe the check for MSYS Makefiles in PR #3838 should be changed to also use a new compiler variable (as I also suggested there).

Reply:
@derekbruening 

I think this code dependent on command shell (bash or smth else)

Reply:
@derekbruening 

If I see that the same conditions will be repeated, I will do so. But for now, unfortunately, this is not obvious. And refactoring can lead to bugs in already working environments.

Reply:
@derekbruening 

Now I don't see any difference between

```
"${CMAKE_GENERATOR}" MATCHES "Visual Studio"
```
(this expression is already present many times in current codebase) and

```
"${CMAKE_GENERATOR}" MATCHES "MSYS Makefiles"
```

Reply:
> I think this code dependent on command shell (bash or smth else)

?  No, this has nothing to do with the shell. This code is locating the preprocessor and setting what flags it takes.  This is only about the *compiler*.

Reply:
> @derekbruening
> 
> Now I don't see any difference between
> 
> ```
> "${CMAKE_GENERATOR}" MATCHES "Visual Studio"
> ```
> 
> (this expression is already present many times in current codebase) and
> 
> ```
> "${CMAKE_GENERATOR}" MATCHES "MSYS Makefiles"
> ```

The difference is that the Visual Studio generator is a unique generator with very different behavior from all the others: it has multiple configurations (debug, release, etc.) supported *at the same time*; it has different parallel build correctness requirements; it does not set up its paths well and we have to augment it.  This "MSYS Makefiles" generator is pretty much identical to "UNIX Makefiles" or "MinGW Makefiles" or "Ninja".  This code here is not limited to "MSYS Makefiles".  It has nothing to do with MSYS.  It only depends on using the gcc (or clang, since the interface is compatible) preprocessor.  That is what this conditional should check.  And there will be many more checks like this that depend only on using a gcc interface.

Reply:
@derekbruening OK, I refactored this code now.